### PR TITLE
[Test] Use UnsafePlainActionFuture for addRentionLease response

### LINKED ISSUE #109163: [CI] LocalIndexFollowingIT testRemoveRemoteConnection failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/mjgqjz3uki7kk/tests/:x-pack:plugin:ccr:internalClusterTest/org.elasticsearch.xpack.ccr.LocalIndexFollowingIT/testRemoveRemoteConnection

**Reproduction line:**
```
./gradlew ':x-pack:plugin:ccr:internalClusterTest' --tests "org.elasticsearch.xpack.ccr.LocalIndexFollowingIT.testRemoveRemoteConnection" -Dtests.seed=ED71ADACFA72AF4F -Dtests.locale=en-MT -Dtests.timezone=Asia/Muscat -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.ccr.LocalIndexFollowingIT#testRemoveRemoteConnection`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testRemoveRemoteConnection'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.ccr.LocalIndexFollowingIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: timed out waiting for shards to initialize

  at org.junit.Assert.fail(Assert.java:89)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.junit.Assert.assertFalse(Assert.java:65)
  at org.elasticsearch.test.ESSingleNodeTestCase.ensureNoInitializingShards(ESSingleNodeTestCase.java:455)
  at org.elasticsearch.test.ESSingleNodeTestCase.tearDown(ESSingleNodeTestCase.java:132)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:1004)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1570)

```

### PR DESCRIPTION
**Pull request by ywangd**
When a transport action fails, it can complete the listener without using the specified executor (#107439). I believe this is what happend in the test failure #109163. The remote is removed before the CCR metaddata is removed. A concurrent add retention lease action can encounter exception which is then completed inline with the same generic thread pool.

Resolves: #109163
Relates: #107439

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java [hunk 1]
```diff
@@ -413,7 +413,10 @@ private boolean assertCompleteAllowed() {
                 + " with waiter on thread "
                 + waiter
                 + ", could deadlock if pool was full\n"
-                + ExceptionsHelper.formatStackTrace(waiter.getStackTrace());
+                + "waiter stacktrace\n"
+                + ExceptionsHelper.formatStackTrace(waiter.getStackTrace())
+                + "\ncompleter stacktrace\n"
+                + ExceptionsHelper.formatStackTrace(Thread.currentThread().getStackTrace());
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java:419]: This should be output by the assertion's stack trace, it was omitted on purpose. Is there a case where we are not dumping this?

**Comment by ywangd** [comment]
A bit more context for why I think this PR should fix the test failure. The failure's [stacktrace](https://gradle-enterprise.elastic.co/s/mjgqjz3uki7kk/tests/task/:x-pack:plugin:ccr:internalClusterTest/details/org.elasticsearch.xpack.ccr.LocalIndexFollowingIT/testRemoveRemoteConnection?focused-execution=1&top-execution=1#L60-L68) basically has the following things in order

1. The test ends
2. The [removeLocalRemote](https://github.com/elastic/elasticsearch/blob/191cacae4855eed6eb172bb29fe903ba83e924d7/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/CcrSingleNodeTestCase.java#L86) method runs as `@After` which disconnects the remote cluster connection
3. At the same time, addRentionLease action failed with the same thread assertion for `generic` threadpool. 
4. CCR auto-follow is removed from the cluster state as another `@After` method [purgeCCRMetadata](https://github.com/elastic/elasticsearch/blob/191cacae4855eed6eb172bb29fe903ba83e924d7/x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/CcrSingleNodeTestCase.java#L80)

The execution of the two `@After` methods are undefined. Locally `purgeCCRMetadata` always runs first for me which should avoid this failure. But apparently they can run in a different order on CI.

I considered whehther the fix should be ensure `purgeCCRMetadata` always runs first. But I decide not to do so because the test internally actively calls `removeLocalRemote` which can potentially run into the same issue. It seems we also do not mandate CCR metadata to be removed before removing the remote cluster. So I don't want to force it in test either.

File: server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java:419]: It *is* included in the specific case [here](https://gradle-enterprise.elastic.co/s/mjgqjz3uki7kk/tests/task/:x-pack:plugin:ccr:internalClusterTest/details/org.elasticsearch.xpack.ccr.LocalIndexFollowingIT/testRemoveRemoteConnection?focused-execution=1&top-execution=1#L103)

I suggest to revert this part.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrRetentionLeases.java [hunk 2]
```diff
@@ -79,7 +81,10 @@ public static Optional<RetentionLeaseAlreadyExistsException> syncAddRetentionLea
         final TimeValue timeout
     ) {
         try {
-            final PlainActionFuture<ActionResponse.Empty> response = new PlainActionFuture<>();
+            final PlainActionFuture<ActionResponse.Empty> response = new UnsafePlainActionFuture<>(
+                Ccr.CCR_THREAD_POOL_NAME,
+                ThreadPool.Names.GENERIC
```
Comment (henningandersen) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrRetentionLeases.java:86]: I believe we need only GENERIC here?

```suggestion
                ThreadPool.Names.GENERIC
```

**Review by henningandersen — COMMENTED** [review]
Left a couple comments.

File: server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java (hunk 1 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java]: You are right. I reverted it in b11a09ac957c88a1bc4fd193278c2ca8be7ab7a6

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrRetentionLeases.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/CcrRetentionLeases.java]: Looking closely these names are for the waiter thread. So you are right. Only `generic` is needed. See b11a09ac957c88a1bc4fd193278c2ca8be7ab7a6

**Review by henningandersen — APPROVED** [review]
LGTM.

**Pull request merged**